core/src/main/scala/com/stripe/dagon/Literal.scala

```diff
@@ -10,15 +10,68 @@ sealed trait Literal[N[_], T] {
 
 object Literal {
 
-  case class Const[N[_], T](override val evaluate: N[T]) extends Literal[N, T]
+  case class Const[N[_], T](override val evaluate: N[T]) extends Literal[N, T] {
+    // cache hashCode since we hit it so much.
+    override val hashCode = (getClass, evaluate).hashCode
 
-  case class Unary[N[_], T1, T2](arg: Literal[N, T1], fn: N[T1] => N[T2]) extends Literal[N, T2]
+    // We often will cache the N[T] -> Literal[N, T] mapping, so we
+    // deal with referentially equal things, which can give cheap equality
+    override def equals(that: Any) =
+      that match {
+        case thatAR: AnyRef if thatAR eq this => true
+        case Const(thatE) => evaluate == thatE
+        case _ => false
+      }
+  }
+
+  case class Unary[N[_], T1, T2](arg: Literal[N, T1], fn: N[T1] => N[T2]) extends Literal[N, T2] {
+    // cache hashCode since we hit it so much.
+    override val hashCode = (getClass, arg, fn).hashCode
```

Comment:
Is there a reason to prefer this approach over `MurmurHash3.productHash(this)`, which I feel like I've seen more often?

Reply:
the only difference is that I am hashing the full class and productHash just hashes the string of the prefix of the name.

Additionally, I avoid an import with this. If you really want the productHash, I can do it.